diff -u nss_db-2.2.3pre1/src/db-open.c.filecopy-leak nss_db-2.2.3pre1/src/db-open.c
--- nss_db-2.2.3pre1/src/db-open.c.filecopy-leak	2014-03-17 10:51:33.084334683 -0300
+++ nss_db-2.2.3pre1/src/db-open.c	2014-03-17 10:51:33.097334683 -0300
@@ -63,8 +63,11 @@
   if (err != 0)
     goto fail;
   filecopy = strdup(file);
+  if (filecopy == NULL)
+    goto fail;
   home = dirname(filecopy);
   err = (dbenv->open)(dbenv, home, DB_INIT_MPOOL | DB_CREATE | DB_PRIVATE, 0);
+  free(filecopy);
   if (err != 0)
     goto fail_env;
   err = db_open (file, DB_BTREE, DB_RDONLY, 0, dbenv, NULL, &db);
@@ -87,7 +90,6 @@
  fail_env:
   dbenv->close (dbenv, 0);
  fail:
-  if (filecopy) free(filecopy);
   if (err > 0)
     errno = err;
   return NSS_STATUS_UNAVAIL;
